### DISCUSSION TITLE
Missing type inference using TypeScript and Immer following example from Readme

### DISCUSSION THREAD

**Discussion by Duckling92** [pmndrs/zustand#2783] [Bug report] [answered]
### Description
When using Immer with Zustand, TypeScript does not catch type errors when setting nested state. The code follows the example from the Zustand documentation, but TypeScript doesn’t detect the type mismatch (state becomes `any`). 

```ts
import { produce } from 'immer'
import { create } from 'zustand'

interface MyStore {
  aNumber: number
  setNumber: () => void
}
const useStore = create<MyStore>()((set) => ({
  aNumber: 0,
  setNumber: () => set(
    produce((state) => {
      state.aNumber = "not a number"
    }),
  ),    
}))
```

### Steps to Reproduce
1. Clone [this](https://github.com/Duckling92/zustand-immer-ts-bug) minimally working example
2. run `npm install`
3. run `npm typecheck`

### Expected behaviour
TypeScript should detect that "not a number" is a string and raise a type error since the aNumber property is typed as a number.

### Actual behaviour
TypeScript does not raise any errors, allowing the assignment of a string to a number field without warnings.

Is this intended behaviour?

### Reproduction Link

https://github.com/Duckling92/zustand-immer-ts-bug

**Comment by dbritto-dev** [comment, marked as answer]
@Duckling92 you are using `produce` from `immer` directly so you need to take care of typing `produce`, that's not zustand issue 

```ts
import { produce } from 'immer'
import { create } from 'zustand'

interface MyStore {
  aNumber: number
  setNumber: () => void
}
const useStore = create<MyStore>()((set, get) => ({
  aNumber: 0,
  setNumber: () => set(
    produce<ReturnType<typeof get>>((state) => {
      state.aNumber = "not a number"
    }),
  ),
}))
```

**Reply by Duckling92** [reply]
Of course, that makes perfect sense. Thank you.
